plugins/statistics.koplugin/main.lua

```diff
@@ -73,58 +86,429 @@ function ReaderStatistics:init()
+                    text =_("New version of statistics plugin is detected.\n" ..
+                        "A new database has been created.")})
+            end
+        end
+        self:saveSettings()
+    end
+    conn:close()
+end
+
+function ReaderStatistics:md5(file)
+    if file == nil then
+        return nil
+    end
+    local md5 = require("ffi/MD5")
+    local lshift = bit.lshift
+    local step, size = 1024, 1024
+    local m = md5.new()
+    local file_handle = io.open(file, 'rb')
+    for i = -1, 10 do
+        file_handle:seek("set", lshift(step, 2*i))
+        local sample = file_handle:read(size)
+        if sample then
+            m:update(sample)
+        else
+            break
+        end
+    end
+    return m:sum()
+end
+
+function ReaderStatistics:createDB(conn)
+    local sql_stmt = [[
+        CREATE TABLE IF NOT EXISTS book
+            (
+                id integer PRIMARY KEY autoincrement,
+                title text,
+                authors text,
+                notes      integer,
+                last_open  integer,
+                highlights integer,
+                pages      integer,
+                series text,
+                language text,
+                md5 text,
+                total_read_time  integer,
+                total_read_pages integer
+            );
+        CREATE TABLE IF NOT EXISTS page_stat
+            (
+                id_book    integer,
+                page       integer NOT NULL,
+                start_time integer NOT NULL,
+                period     integer NOT NULL,
+                UNIQUE (page, start_time),
+                FOREIGN KEY(id_book) REFERENCES book(id)
+             );
+        CREATE TABLE IF NOT EXISTS info
+             (
+                 version integer
+             );
+    ]]
+    conn:exec(sql_stmt)
+    --DB structure version - now is version 1
+    local stmt = conn:prepare("INSERT INTO info values (?)")
+    stmt:reset():bind("1"):step()
+    stmt:close()
+end
+
+function ReaderStatistics:addBookStatToDB(book_stats, conn)
+    local res
+    local id_book
+    local last_open_book = 0
+    local start_open_page
+    local diff_time
+    local total_read_pages = 0
+    local total_read_time = 0
+    local sql_stmt
+    if book_stats.total_time_in_sec and book_stats.total_time_in_sec > 0
+        and util.tableSize(book_stats.performance_in_pages) > 0 then
+        local read_pages = util.tableSize(book_stats.performance_in_pages)
+        logger.dbg("Insert to database: " .. book_stats.title)
+        local book_title = book_stats.title:gsub("'", "''")
+        local book_authors = book_stats.authors:gsub("'", "''")
+        sql_stmt = [[
+            SELECT count(id)
+            FROM   book
+            WHERE  title = '%s'
+                AND    authors = '%s'
+                AND    md5 = '%s';
+        ]]
+        res = conn:rowexec(string.format(sql_stmt, book_title, book_authors, self:md5(book_stats.file)))
```

Comment:
For dynamic queries, always use prepare then bind to execute it :) Doing manual string formatting is error prone because there are all kinds of weird keywords that need to be escaped to construct a valid query. Doing our own escaping would eventually end up with a reimplementation of bind, so might as well use bind from the start. bind should automatically escape the argument for you based on Lua types, so you won't need to do the tile and author gsub.